### DISCUSSION TITLE
Provide reason with rules

### DISCUSSION THREAD

**Discussion by SISheogorath** [ansible/ansible-lint#936] [Ideas]
##### Summary

Provide a reason along the side with the explanation how to solve the linting error in order to allow users to make informed decisions about the linting error.

##### Issue Type

- Feature Idea

##### Additional Information

Currently the default rules simply state what is wrong and how to fix it. What they don't tell is why it's wrong or simply worth checking. For example: `403: Package installs should not use latest` simply states that `Package installs should use state=present with or without a version` but it doesn't tell why. This leaves me as a user in the position: Should I follow the linters advice or simply add an exception for the rule? If it would tell `Package installs should use state=present with or without a version because it can cause instability in the setup` or `Package installs should use state=present with or without a version because state=latest can harm makes the role no longer idempotent`.

I guess I simply miss a "because" section for the default rules and it would be great to get it, as it would help to transfer knowledge and provide the ability to make better informed decisions on why a linter rule exists and whether something is intended or unintended behaviour.

**Comment by bendem** [comment]
This. Linting allows you to discover things you do wrong, if there is no reasoning, you can't learn why it's wrong, you can't make a decision, it becomes cargo-culting.

**Comment by matkoniecz** [comment]
It would be very useful. For example I got

"Package installs should not use latest"

on role that have as whole point to upgrade `pip` and `youtube-dl`.

I am not sure what is supposed to be wrong with them. If the reason is "it can cause instability in the setup"  then I would happily ignore it - as in this case whole point is to introduce instability (AKA upgrade them as needed).

But if `latest` should not be used because it is buggy, misnamed or something then I would care about it.

**Comment by ssbarnea** [comment]
This looks like a good idea, a pull-requests would be welcomed.

**Comment by widhalmt** [comment]
Maybe this overloads the output of `ansible-lint`. I think the online documentation of rules is really enough. The only improvement I can think of is a more prominent link or tip how to find them online.
